ui/src/BaseEditor.ts

```diff
@@ -0,0 +1,169 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as fse from 'fs-extra';
+import * as path from 'path';
+import * as vscode from 'vscode';
+import { DialogResponses } from './DialogResponses' ;
+import { UserCancelledError } from './errors';
+import { localize } from "./localize";
+import { TemporaryFile } from './utils/TemporaryFile';
+
+// tslint:disable-next-line:no-unsafe-any
+export abstract class BaseEditor<ContextT> implements vscode.Disposable {
+    private fileMap: { [key: string]: [vscode.TextDocument, ContextT] } = {};
+    private ignoreSave: boolean = false;
+
+    constructor(readonly showSavePromptKey: string, readonly outputChanel?: vscode.OutputChannel) {
+    }
+
+    public abstract getData(context: ContextT): Promise<string>;
+    public abstract updateData(context: ContextT, data: string): Promise<string>;
+    public abstract getFilename(context: ContextT): Promise<string>;
+    public abstract getSize(context: ContextT): Promise<number>;
+    public abstract getSaveConfirmationText(context: ContextT): Promise<string>;
+
+    public async showEditor(context: ContextT): Promise<void> {
+        const size: number = await this.getSize(context);
+        const fileName: string = await this.getFilename(context);
+        const splitFileName: string[] = fileName.split('.');
+        const extension: string = splitFileName[splitFileName.length - 1];
+
+        this.appendToOutput(localize('opening', 'Opening "{0}" ...', fileName));
+        if (size > 50 /*Megabytes*/) {
+            const message: string = localize('tooLargeError', '"{0}" is too large to download.', fileName);
+
+            await vscode.window.showWarningMessage(message, DialogResponses.OK);
+            this.appendLineToOutput(localize('failed', "Failed."));
+            this.appendLineToOutput(localize('errorDetails', 'Error Details: {0}', message));
+        } else if (extension === 'exe' || extension === 'img' || extension === 'zip') {
+            const message: string = localize('unsupportedError', '"{0}" has an unsupported file extension.', fileName);
+
+            await vscode.window.showWarningMessage(message, DialogResponses.OK);

```

Comment:
I would throw an error rather than show a message for two reasons:
1. If you swallow the error like this, it's never tracked in telemetry
1. Our extensions already have logic to display errors to the user. We should just leverage that to be consistent

Reply:
Good point. Should we create custom errors for this or would is this okay for a generic type?

Reply:
Generic is fine IMO